README
PPIx::Element::Sub

This module contains a handful of utility functions for finding logical
"PPI" nodes in the relative proximity of any given "element".

It contains three primary exportable utility functions:

"identify_sub" - Find a logically enclosing sub
"identify_next_sub" - Find a logically succeeding sub
"identify_associated_sub" - Find a logically enclosing sub, or if none,
a logically succeeding sub.

INSTALLATION

This is a Perl module distribution. It should be installed with whichever
tool you use to manage your installation of Perl, e.g. any of

  cpanm .
  cpan  .
  cpanp -i .

Consult http://www.cpan.org/modules/INSTALL.html for further instruction.
Should you wish to install this module manually, the procedure is

  perl Makefile.PL
  make
  make test
  make install

COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric
<kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Release history for PPIx-Element-Sub

0.001001 2015-10-19T14:50:31Z 624af94

0.001000 2015-10-19T14:48:49Z 7e16923
 - First version.


t/lib/SubIs.pm
use strict;
use warnings;

package SubIs;

use Exporter 5.57 qw( import );
use Test::Builder;
use PPIx::Element::Sub qw( identify_sub identify_next_sub identify_associated_sub );

our @EXPORT = qw( sub_is next_sub_is assoc_sub_is );

sub sub_is($$$) {
    my ( $element, $name, $reason ) = @_;
    my $builder    = Test::Builder->new();
    my $identified = identify_sub($element);
    if ( not defined $name ) {
        return 1 if $builder->is_eq( $identified, undef, $reason );
        $builder->diag("undef expected, got result");
    }
    else {

        return 1
          if $builder->ok( defined $identified, "$reason: got an object" )
          and $builder->is_eq( $identified->name, $name, "$reason: name matches" );
    }
    $builder->diag( "element class:" . $element->class );
    $builder->diag( "element content:" . $element->content );
    if ( defined $identified ) {
        $builder->diag( "got result:" . $identified->class );
        $builder->diag( "got content:" . $identified->content );
    }
}

sub next_sub_is($$$) {
    my ( $element, $name, $reason ) = @_;
    my $builder    = Test::Builder->new();
    my $identified = identify_next_sub($element);
    if ( not defined $name ) {
        return 1 if $builder->ok( !defined $identified, $reason );
        $builder->diag("undef expected, got result");
    }
    else {
        return 1
          if $builder->ok( defined $identified, "$reason: got an object" )
          and $builder->is_eq( $identified->name, $name, "$reason: name matches" );
    }
    $builder->diag( "element class:" . $element->class );
    $builder->diag( "element content:" . $element->content );
    if ( defined $identified ) {
        $builder->diag( "got result:" . $identified->class );
        $builder->diag( "got content:" . $identified->content );
    }
}

sub assoc_sub_is($$$) {
    my ( $element, $name, $reason ) = @_;
    my $builder    = Test::Builder->new();
    my $identified = identify_associated_sub($element);
    if ( not defined $name ) {
        return 1 if $builder->ok( !defined $identified, $reason );
        $builder->diag("undef expected, got result");
    }
    else {
        return 1
          if $builder->ok( defined $identified, "$reason: got an object" )
          and $builder->is_eq( $identified->name, $name, "$reason: name matches" );
    }
    $builder->diag( "element class:" . $element->class );
    $builder->diag( "element content:" . $element->content );
    if ( defined $identified ) {
        $builder->diag( "got result:" . $identified->class );
        $builder->diag( "got content:" . $identified->content );
    }
}



t/corpus/noscope.pm
# no subs or packages in this file


t/corpus/aftersub.pm
sub foo {

}

# this comment is outside the sub and not before it


lib/PPIx/Element/Sub.pm
use 5.006;    # our
use strict;
use warnings;

package PPIx::Element::Sub;

our $VERSION = '0.001001'; # TRIAL

# ABSTRACT: Find subroutines associated with any element.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

our @EXPORT_OK = qw( identify_sub identify_next_sub identify_associated_sub );

use Exporter 5.57 qw( import );
use Scalar::Util qw( refaddr );

sub identify_sub {
    my ($element) = @_;

    # any <sub Foo;> contains itself.
    return $element if $element->isa('PPI::Statement::Sub');

    # no parents => can't find a sub => must be main execution
    return unless $element->can('parent') and defined( my $parent = $element->parent );

    # If we have a parent, ask it for which sub we're in.
    return identify_sub($parent);
}

sub identify_next_sub {
    my ($element) = @_;

    # Parentless nodes can't have siblings
    return unless $element->can('parent') and defined( my $parent = $element->parent );

    # parents without children can't give us siblings
    # but *their* parents could
    return identify_next_sub($parent) unless $parent->can('children');

    my $own_addr  = refaddr($element);
    my $seen_self = 0;

    # Search direct siblings deeply
    for my $sibling ( $parent->children ) {

        # Iterate forwards until self is found.
        if ( not $seen_self ) {
            $seen_self = 1 if refaddr($sibling) eq $own_addr;
            next;
        }

        # For successive siblings
        # return the first sub found
        return $sibling if $sibling->isa('PPI::Statement::Sub');

        # If a sibling has children, a sub could be one of them
        # note: a false value == fail, but explit undef is invalid query
        if ( $sibling->can('find_first') and my $result = $sibling->find_first('PPI::Statement::Sub') ) {
            return $result;
        }
    }

    # When no sibling is a sub, assume that the sub is further down
    # the document, making it a sibling that succeeds a parent, or a descendent
    # of a parents sibling.
    return identify_next_sub($parent);
}

sub identify_associated_sub {
    my ($element) = @_;

    # First, resolve to the containers, as all comments inside a
    # sub are related to the sub itself
    if ( my $parent = identify_sub($element) ) {
        return $parent;
    }

    # Otherwise, look into the next-sibling tree to find the subsequent sub
    return identify_next_sub($element);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

PPIx::Element::Sub - Find subroutines associated with any element.

=head1 VERSION

version 0.001001

=head1 SYNOPSIS

  use PPIx::Element::Sub qw( identify_associated_sub );

  my $assoc_sub = identify_associated_sub( $comment_element );

=head1 DESCRIPTION

This module contains a handful of utility functions for finding logical C<PPI>
nodes in the relative proximity of any given C<element>.

It contains three primary exportable utility functions:

=over 4

=item C<identify_sub> - Find a logically enclosing sub

=item C<identify_next_sub> - Find a logically succeeding sub

=item C<identify_associated_sub> - Find a logically enclosing sub, or if none,
a logically succeeding sub.

=back

=head1 FUNCTIONS

=head2 identify_sub

Given an C<element>, ascertain the logical C<sub> that contains it.

  my $sub = identify_sub( $element )

Due to structure of C<PPI> documents, this is simply the nearest C<parent> of
C<element> that is a C<sub>, which could be the C<element> itself.

  package Foo;  # undef
                # ↑
  sub           # sub 'bar'
  bar           # ↑
  {             # ↑
                # ↑
  }             # ↑
                # undef

Returns C<undef> if no C<sub> parent can be derived.

=head2 identify_next_sub

Given an C<element>, ascertain the I<next> C<sub> that appears logically
I<after> it in the document.

  my $sub = identify_next_sub( $element );

Returns C<undef> if no C<next> <sub> can be found.

  package Foo;  # sub 'bar'
                # ↑
  sub           # quux
  bar           # ↑
  {             # ↑
                # ↑
  }             # ↑
                # ↑
  sub quux      # undef
  {             # ↑
  }             # ↑

=head2 identify_associated_sub

This is a combination of C<identify_sub> and C<identify_next_sub> aimed at
comments, as logically, any comment inside a C<sub> pertains to the C<sub>
itself, but any comment B<before> a C<sub> may also pertain to that C<sub>.

  my $sub = identify_associated_sub( $element );

Any C<element> inside a C<sub> is associated with the C<sub> itself.

Any C<element> I<before> a C<sub> is associated with the subsequent C<sub>

  package Foo;  # sub 'bar'
                # ↑
  sub           # ↑
  bar           # ↑
  {             # ↑
                # ↑
  }             # ↑
                # sub 'quux'
  sub quux      # ↑
  {             # ↑
  }             # ↑
                # undef

Returns C<undef> if no C<sub> can be determined.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
## no critic (Modules::RequireVersionVar)

# ABSTRACT: Write an INI file from a bundle

use 5.008;    # utf8
use strict;
use warnings;
use utf8;

our $VERSION = 0.001;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );
use Path::Tiny qw(path);

## no critic (ErrorHandling::RequireUseOfExceptions)
my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw());
for my $wordlist (@stopwords) {
    $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $wordlist ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );

#$bundle->remove_policy('ErrorHandling::RequireUseOfExceptions');
#$bundle->remove_policy('CodeLayout::RequireUseUTF8');
#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;

{
    my $rcfile = path('./perlcritic.rc')->openw_utf8;
    $rcfile->print( $config->as_ini, "\n" );
    close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
    my $target = path('./misc');
    $target->mkpath if not $target->is_dir;

    my $depsfile = $target->child('perlcritic.deps')->openw_utf8;
    for my $key ( sort keys %{$deps} ) {
        $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
        *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
    }
    close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.040.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Find subroutines associated with any element.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "PPIx-Element-Sub",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "PPIx::Element::Sub",
  "PREREQ_PM" => {
    "Exporter" => "5.57",
    "Scalar::Util" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "PPI::Util" => 0,
    "Test::Builder" => 0,
    "Test::More" => "0.89",
    "lib" => 0
  },
  "VERSION" => "0.001001",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t t/internals/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => "5.57",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "PPI::Util" => 0,
  "Scalar::Util" => 0,
  "Test::Builder" => 0,
  "Test::More" => "0.89",
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
